pallets/poa/src/lib.rs

```diff
@@ -0,0 +1,710 @@
+                fail!(Error::<T>::AlreadyQueuedForRemoval)
+            }
+        } else {
+            removals.insert(&validator_id);
+            false
+        };
+
+        let active_validators: Vec<T::AccountId> = Self::active_validators().into();
+        let validators_to_add: Vec<T::AccountId> = Self::validators_to_add().into();
+
+        // Construct a set of potential validators and don't allow all the potential validators
+        // to be removed as that will prevent the node from starting.
+        // This takes away the ability of do a remove before an add but should not matter.
+        let mut potential_new_vals = BTreeSet::new();
+        for v in active_validators.iter() {
+            potential_new_vals.insert(v);
+        }
+        for v in validators_to_add.iter() {
+            potential_new_vals.insert(v);
+        }
+
+        // There should be at least 1 id in potential_new_vals that is not in removals
+        let diff: Vec<_> = potential_new_vals.difference(&removals).collect();
+        if diff.is_empty() {
+            print("Cannot remove. Need at least 1 active validator");
+            fail!(Error::<T>::NeedAtLeast1Validator)
+        }
+
+        // Add validator
+        if !already_queued_for_rem {
+            validators_to_remove.push(validator_id.clone());
+            <RemoveValidators<T>>::put(validators_to_remove);
+            Self::deposit_event(RawEvent::ValidatorRemoved(validator_id));
+        }
+
+        if short_circuit {
+            Self::short_circuit_current_epoch();
+        }
+        Ok(())
+    }
+
+    fn swap_validator_(
+        old_validator_id: T::AccountId,
+        new_validator_id: T::AccountId,
+    ) -> dispatch::DispatchResult {
+        let active_validators: Vec<T::AccountId> = Self::active_validators().into();
+
+        let mut found = false;
+        for v in active_validators.iter() {
+            if *v == new_validator_id {
+                print("New validator to swap in already present");
+                fail!(Error::<T>::SwapInFailed)
+            }
+            if *v == old_validator_id {
+                found = true;
+            }
+        }
+
+        if !found {
+            print("Validator to swap out not present");
+            fail!(Error::<T>::SwapOutFailed)
+        }
+
+        <HotSwap<T>>::put((old_validator_id, new_validator_id));
+        Ok(())
+    }
+
+    /// Takes a validator id and a mutable vector of validator ids and remove any occurrence from
+    /// the mutable vector. Returns number of removed occurrences
+    fn remove_validator_id(id: &T::AccountId, validators: &mut Vec<T::AccountId>) -> usize {
+        let old_size = validators.len();
+        validators.retain(|v| v != id);
+        old_size - validators.len()
+    }
+
+    /// Update active validator set if needed and return if the active validator set changed and the
+    /// count of the new active validators.
+    fn update_active_validators_if_needed() -> (bool, u8) {
```

Comment:
btw, I don't recommend breaking this one up into smaller functions. I think this function can be made unnecessary by changing other aspects of the module. For example, changing `ActiveValidators` to `PlannedValidators` would allow queue, removal, and swap processing to be done within `add_validator`, `remove_validator`, and `swap_validators` respectively.

Reply:
Can you describe the change? Does that mean manipulating session pallet's storage directly?

Reply:
`ActiveValidators` would be changed to `PlannedValidators` to reflect it's new meaning. `PlannedValidators` would not be interpreted as the *current* validator set, but the validator set that is planned for a future epoch/session. Changing the semantics of `PlannedValidators` in this way allows it to be mutated directly instead of storing `remove` and `hot_swap` operations to be processed later.

I don't think the `pallet_session::Validators` is intended to be written to by other modules. I could be wrong.

Reply:
See my answer to comment above

Reply:
Do you mean to say that if you use `pallet_session::Validators` as a source of truth, you will need to call `rotate_session` more often?

Reply:
tags:
- best-practices
  1. avoid long functions, deep nesting, and high [cyclomatic complexity](https://en.wikipedia.org/wiki/Cyclomatic_complexity)
  2. beware the pitfall of premature optimization
- readability-surface
  1. this function took a me while to grok